stan/math/gpu/err/check_matrix_gpu.hpp

```diff
@@ -0,0 +1,184 @@
+                                 &nan_flag);
+
+    kernel_check_nan.setArg(0, y.buffer());
+    kernel_check_nan.setArg(1, y.rows());
+    kernel_check_nan.setArg(2, y.cols());
+    kernel_check_nan.setArg(3, buffer_nan_flag);
+
+    cmd_queue.enqueueNDRangeKernel(kernel_check_nan, cl::NullRange,
+                                   cl::NDRange(y.rows(), y.cols()),
+                                   cl::NullRange);
+
+    cmd_queue.enqueueReadBuffer(buffer_nan_flag, CL_TRUE, 0, sizeof(int),
+                                &nan_flag);
+    //  if NaN values were found in the matrix
+    if (nan_flag) {
+      domain_error(function, name, "has NaN values", "");
+    }
+  } catch (const cl::Error& e) {
+    check_opencl_error("nan_check", e);
+  }
+}
+/**
+ * Check if the <code>matrix_gpu</code> is symmetric
+ *
+ * @param function Function name (for error messages)
+ * @param name Variable name (for error messages)
+ * @param y <code>matrix_gpu</code> to test
+ *
+ * @throw <code>std::domain_error</code> if
+ *    the matrix is not symmetric.
+ */
+inline void check_symmetric(const char* function, const char* name,
+                            const matrix_gpu& y) {
+  if (y.size() == 0)
+    return;
+  check_square(function, name, y);
+  cl::Kernel kernel_check_symmetric
+      = opencl_context.get_kernel("check_symmetric");
+  cl::CommandQueue cmd_queue = opencl_context.queue();
+  cl::Context& ctx = opencl_context.context();
+  try {
+    int symmetric_flag = 0;
+    cl::Buffer buffer_symmetric_flag(ctx, CL_MEM_READ_WRITE, sizeof(int));
+    cmd_queue.enqueueWriteBuffer(buffer_symmetric_flag, CL_TRUE, 0, sizeof(int),
+                                 &symmetric_flag);
+
+    kernel_check_symmetric.setArg(0, y.buffer());
+    kernel_check_symmetric.setArg(1, y.rows());
+    kernel_check_symmetric.setArg(2, y.cols());
+    kernel_check_symmetric.setArg(3, buffer_symmetric_flag);
+    kernel_check_symmetric.setArg(4, math::CONSTRAINT_TOLERANCE);
+
+    cmd_queue.enqueueNDRangeKernel(kernel_check_symmetric, cl::NullRange,
+                                   cl::NDRange(y.rows(), y.cols()),
+                                   cl::NullRange);
+
+    cmd_queue.enqueueReadBuffer(buffer_symmetric_flag, CL_TRUE, 0, sizeof(int),
+                                &symmetric_flag);
+    //  if the matrix is not symmetric
+    if (symmetric_flag) {
+      domain_error(function, name, "is not symmetric", "");
+    }
+  } catch (const cl::Error& e) {
+    check_opencl_error("symmetric_check", e);
+  }
+}
+/**
+ * Check if the <code>matrix_gpu</code> has zeros on the diagonal
+ *
+ * @param function Function name (for error messages)
+ * @param name Variable name (for error messages)
+ * @param y <code>matrix_gpu</code> to test
+ *
+ * @throw <code>std::domain_error</code> if
+ *    any diagonal element of the matrix is zero.
+ */
+inline void check_diagonal_zeros(const char* function, const char* name,
```

Comment:
Tests for this function? Does this actually do the right thing for checking all diagonals?

Reply:
Added a test.